docs/reference/developer/adr/0017-restrict-dataset-to-items-with-matching-dimensionality.rst

```diff
@@ -0,0 +1,98 @@
+ADR 0017: Restrict Dataset to items with matching dimensionality
+================================================================
+
+- Status: proposed
+- Deciders: Jan-Lukas, Neil, Simon, Sunyoung
+- Date: 2023-07-04
+
+Context
+-------
+
+Background
+~~~~~~~~~~
+
+``Dataset`` was conceived as a way to represent a collection of items with compatible dimension extents, but possibly different sets or subsets of dimensions.
+Recently we have introduced ``DataGroup``, which drops the restriction of compatible dimension extents but, unlike ``Dataset``, does in turn not provide support for joint coordinates.
+The addition of ``DataGroup`` was triggered by a long series of cases where we realized that ``Dataset`` is not useful and flexbile enough in real applications.
+This is not to say that ``Dataset`` is entirely useless, but it is not as useful as we had hoped.
+One key area that is not covered by ``DataGroup`` is the representation of table-like data (or multi-dimensional generalizations thereof), in a manner similar to ``pandas.DataFrame``.
+
+The recent decision to remove support for ``attrs``, including addition of an ``aligned`` flag, has made the situation worse, as this added a number of edge cases where the semantics of ``Dataset`` are not clear.
+
+Finally, there have been discussions around structure-of-array data-types.
+``Dataset`` could be used for this purpose, but the current semantics may be too complex for comfort.
+
+Analysis
+~~~~~~~~
+
+There are two possible ways of reasoning about ``Dataset``.
+Firstly, we may argue that while technically complex, the work has already been done, and the problems detailed below are encountered only in edge cases.
```

Comment:
Are we finally going to refactor the C++ implementation of `DataArray` to not use `Dataset` but its own dedicated implementation? (at the moment, it almost feels like we already have legacy code in the C++ that no one wants to touch because it 'works')
(and maybe the same with the 'buckets' which should really be called 'bins'?)

Reply:
We already did that several years ago.